backends/cuda/ceed-cuda-vector.c

```diff
@@ -54,46 +71,149 @@ static inline int CeedVectorSyncD2H_Cuda(const CeedVector vec) {
   int ierr;
   Ceed ceed;
   ierr = CeedVectorGetCeed(vec, &ceed); CeedChkBackend(ierr);
-  CeedVector_Cuda *data;
-  ierr = CeedVectorGetData(vec, &data); CeedChkBackend(ierr);
+  CeedVector_Cuda *impl;
+  ierr = CeedVectorGetData(vec, &impl); CeedChkBackend(ierr);
 
-  ierr = cudaMemcpy(data->h_array, data->d_array, bytes(vec),
+  if (!impl->d_array)
+    // LCOV_EXCL_START
+    return CeedError(ceed, CEED_ERROR_BACKEND,
+                     "No valid device data to sync to host");
+  // LCOV_EXCL_STOP
+
+  if (impl->h_array_borrowed) {
+    impl->h_array = impl->h_array_borrowed;
+  } else if (impl->h_array_owned) {
+    impl->h_array = impl->h_array_owned;
+  } else {
+    CeedInt length;
+    ierr = CeedVectorGetLength(vec, &length); CeedChkBackend(ierr);
+    ierr = CeedCalloc(length, &impl->h_array_owned);  CeedChkBackend(ierr);
+    impl->h_array = impl->h_array_owned;
+  }
+
+  ierr = cudaMemcpy(impl->h_array, impl->d_array, bytes(vec),
                     cudaMemcpyDeviceToHost); CeedChk_Cu(ceed, ierr);
+
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Sync arrays
+//------------------------------------------------------------------------------
+static inline int CeedVectorSync_Cuda(const CeedVector vec, CeedMemType mtype) {
+  switch (mtype) {
+  case CEED_MEM_HOST: return CeedVectorSyncD2H_Cuda(vec);
+  case CEED_MEM_DEVICE: return CeedVectorSyncH2D_Cuda(vec);
+  }
+  return CEED_ERROR_UNSUPPORTED;
+}
+
+//------------------------------------------------------------------------------
+// Set all pointers as invalid
+//------------------------------------------------------------------------------
+static inline int CeedVectorSetAllInvalid_Cuda(const CeedVector vec) {
+  int ierr;
+  CeedVector_Cuda *impl;
+  ierr = CeedVectorGetData(vec, &impl); CeedChkBackend(ierr);
+
+  impl->h_array = NULL;
+  impl->d_array = NULL;
+
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Check if CeedVector has any valid pointer
+//------------------------------------------------------------------------------
+static inline int CeedVectorHasValidArray_Cuda(const CeedVector vec,
+    bool *has_valid_array) {
+  int ierr;
+  CeedVector_Cuda *impl;
+  ierr = CeedVectorGetData(vec, &impl); CeedChkBackend(ierr);
+
+  *has_valid_array = !!impl->h_array || !!impl->d_array;
+
```

Comment:
Is there a reason for `!!` rather than just checking for `impl->h_array`?

Reply:
Just being explicit that I want a boolean value